[CI] FileServiceAccountTokenStoreTests testAutoReload failing

**Build scan:**
https://gradle-enterprise.elastic.co/s/a6hdquncqz7q4/tests/:x-pack:plugin:security:test/org.elasticsearch.xpack.security.authc.service.FileServiceAccountTokenStoreTests/testAutoReload

**Reproduction line:**
`gradlew ':x-pack:plugin:security:test' --tests "org.elasticsearch.xpack.security.authc.service.FileServiceAccountTokenStoreTests.testAutoReload" -Dtests.seed=91026D8CFE7DC817 -Dtests.locale=hu -Dtests.timezone=Pacific/Saipan -Druntime.java=8`

**Applicable branches:**
7.x

**Reproduces locally?:**
No

**Failure history:**
https://gradle-enterprise.elastic.co/scans/tests?tests.container=org.elasticsearch.xpack.security.authc.service.FileServiceAccountTokenStoreTests&tests.test=testAutoReload

**Failure excerpt:**
```
java.lang.AssertionError: 
Expected: is <true>
     but: was <false>

  at __randomizedtesting.SeedInfo.seed([91026D8CFE7DC817:D7BE7215BB27A801]:0)
  at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
  at org.junit.Assert.assertThat(Assert.java:956)
  at org.junit.Assert.assertThat(Assert.java:923)
  at org.elasticsearch.xpack.security.authc.service.FileServiceAccountTokenStoreTests.testAutoReload(FileServiceAccountTokenStoreTests.java:156)
  at sun.reflect.NativeMethodAccessorImpl.invoke0(NativeMethodAccessorImpl.java:-2)
  at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
  at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
  at java.lang.reflect.Method.invoke(Method.java:498)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1750)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$8.evaluate(RandomizedRunner.java:938)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:974)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:988)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:49)
  at org.apache.lucene.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:45)
  at org.apache.lucene.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:48)
  at org.apache.lucene.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:64)
  at org.apache.lucene.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:47)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:368)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:817)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:468)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:947)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:832)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:883)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:894)
  at org.apache.lucene.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:45)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:41)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)
  at org.apache.lucene.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:47)
  at org.apache.lucene.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:64)
  at org.apache.lucene.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:54)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:368)
  at java.lang.Thread.run(Thread.java:748)

```

--- a/x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/service/FileServiceAccountTokenStoreTests.java
+++ b/x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/service/FileServiceAccountTokenStoreTests.java
@@ -36,11 +36,12 @@ import java.nio.file.StandardOpenOption;
 import java.util.Collection;
 import java.util.List;
 import java.util.Map;
-import java.util.concurrent.CountDownLatch;
 import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
 
 import static org.hamcrest.Matchers.containsString;
 import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.greaterThan;
 import static org.hamcrest.Matchers.is;
 import static org.hamcrest.Matchers.notNullValue;
 import static org.hamcrest.Matchers.nullValue;
@@ -101,10 +102,11 @@ public class FileServiceAccountTokenStoreTests extends ESTestCase {
 
     public void testParseFileNotExists() throws IllegalAccessException, IOException {
         Logger logger = CapturingLogger.newCapturingLogger(Level.TRACE, null);
+        final List<String> events = CapturingLogger.output(logger.getName(), Level.TRACE);
+        events.clear();
         final Map<String, char[]> tokenHashes =
             FileServiceAccountTokenStore.parseFile(getDataPath("service_tokens").getParent().resolve("does-not-exist"), logger);
         assertThat(tokenHashes.isEmpty(), is(true));
-        final List<String> events = CapturingLogger.output(logger.getName(), Level.TRACE);
         assertThat(events.size(), equalTo(2));
         assertThat(events.get(1), containsString("does not exist"));
     }
@@ -115,47 +117,53 @@ public class FileServiceAccountTokenStoreTests extends ESTestCase {
         Files.createDirectories(configDir);
         Path targetFile = configDir.resolve("service_tokens");
         Files.copy(serviceTokensSourceFile, targetFile, StandardCopyOption.REPLACE_EXISTING);
-        final Hasher hasher = Hasher.resolve(settings.get("xpack.security.authc.service_token_hashing.algorithm"));
+        final String hashingAlgo = settings.get("xpack.security.authc.service_token_hashing.algorithm");
+        final Hasher hasher = Hasher.resolve(hashingAlgo);
         try (ResourceWatcherService watcherService = new ResourceWatcherService(settings, threadPool)) {
-            final CountDownLatch latch = new CountDownLatch(5);
+            final AtomicInteger counter = new AtomicInteger(0);
 
             FileServiceAccountTokenStore store = new FileServiceAccountTokenStore(env, watcherService, threadPool,
                 mock(CacheInvalidatorRegistry.class));
-            store.addListener(latch::countDown);
+            store.addListener(counter::getAndIncrement);
             //Token name shares the hashing algorithm name for convenience
-            String tokenName = settings.get("xpack.security.authc.service_token_hashing.algorithm");
-            final String qualifiedTokenName = "elastic/fleet-server/" + tokenName;
+            final String qualifiedTokenName = "elastic/fleet-server/" + hashingAlgo;
             assertThat(store.getTokenHashes().containsKey(qualifiedTokenName), is(true));
 
+            final int oldValue1 = counter.get();
             // A blank line should not trigger update
             try (BufferedWriter writer = Files.newBufferedWriter(targetFile, StandardCharsets.UTF_8, StandardOpenOption.APPEND)) {
                 writer.append("\n");
             }
             watcherService.notifyNow(ResourceWatcherService.Frequency.HIGH);
-            if (latch.getCount() != 5) {
+            if (counter.get() != oldValue1) {
                 fail("Listener should not be called as service tokens are not changed.");
             }
             assertThat(store.getTokenHashes().containsKey(qualifiedTokenName), is(true));
 
             // Add a new entry
+            final int oldValue2 = counter.get();
             final char[] newTokenHash =
                 hasher.hash(new SecureString("46ToAwIHZWxhc3RpYwVmbGVldAZ0b2tlbjEWWkYtQ3dlWlVTZldJX3p5Vk9ySnlSQQAAAAAAAAA".toCharArray()));
             try (BufferedWriter writer = Files.newBufferedWriter(targetFile, StandardCharsets.UTF_8, StandardOpenOption.APPEND)) {
                 writer.newLine();
                 writer.append("elastic/fleet-server/token1:").append(new String(newTokenHash));
             }
-            assertBusy(() -> assertEquals("Waited too long for the updated file to be picked up", 4, latch.getCount()),
-                5, TimeUnit.SECONDS);
-            assertThat(store.getTokenHashes().containsKey("elastic/fleet-server/token1"), is(true));
+            assertBusy(() -> {
+                assertThat("Waited too long for the updated file to be picked up", counter.get(), greaterThan(oldValue2));
+                assertThat(store.getTokenHashes().containsKey("elastic/fleet-server/token1"), is(true));
+            }, 5, TimeUnit.SECONDS);
 
             // Remove the new entry
+            final int oldValue3 = counter.get();
             Files.copy(serviceTokensSourceFile, targetFile, StandardCopyOption.REPLACE_EXISTING);
-            assertBusy(() -> assertEquals("Waited too long for the updated file to be picked up", 3, latch.getCount()),
-                5, TimeUnit.SECONDS);
-            assertThat(store.getTokenHashes().containsKey("elastic/fleet-server/token1"), is(false));
-            assertThat(store.getTokenHashes().containsKey(qualifiedTokenName), is(true));
+            assertBusy(() -> {
+                assertThat("Waited too long for the updated file to be picked up", counter.get(), greaterThan(oldValue3));
+                assertThat(store.getTokenHashes().containsKey("elastic/fleet-server/token1"), is(false));
+                assertThat(store.getTokenHashes().containsKey(qualifiedTokenName), is(true));
+            }, 5, TimeUnit.SECONDS);
 
             // Write a mal-formatted line
+            final int oldValue4 = counter.get();
             if (randomBoolean()) {
                 try (BufferedWriter writer = Files.newBufferedWriter(targetFile, StandardCharsets.UTF_8, StandardOpenOption.APPEND)) {
                     writer.newLine();
@@ -168,24 +176,29 @@ public class FileServiceAccountTokenStoreTests extends ESTestCase {
                     writer.append("elastic/fleet-server/tokenx:").append(new String(newTokenHash));
                 }
             }
-            assertBusy(() -> assertEquals("Waited too long for the updated file to be picked up", 2, latch.getCount()),
-                5, TimeUnit.SECONDS);
-            assertThat(store.getTokenHashes().isEmpty(), is(true));
+            assertBusy(() -> {
+                assertThat("Waited too long for the updated file to be picked up", counter.get(), greaterThan(oldValue4));
+                assertThat(store.getTokenHashes().isEmpty(), is(true));
+            }, 5, TimeUnit.SECONDS);
 
             // Restore to original file again
+            final int oldValue5 = counter.get();
             Files.copy(serviceTokensSourceFile, targetFile, StandardCopyOption.REPLACE_EXISTING);
-            assertBusy(() -> assertEquals("Waited too long for the updated file to be picked up", 1, latch.getCount()),
-                5, TimeUnit.SECONDS);
-            assertThat(store.getTokenHashes().containsKey(qualifiedTokenName), is(true));
+            assertBusy(() -> {
+                assertThat("Waited too long for the updated file to be picked up", counter.get(), greaterThan(oldValue5));
+                assertThat(store.getTokenHashes().containsKey(qualifiedTokenName), is(true));
+            }, 5, TimeUnit.SECONDS);
 
             // Duplicate entry
+            final int oldValue6 = counter.get();
             try (BufferedWriter writer = Files.newBufferedWriter(targetFile, StandardCharsets.UTF_8, StandardOpenOption.APPEND)) {
                 writer.newLine();
-                writer.append(qualifiedTokenName + ":").append(new String(newTokenHash));
+                writer.append(qualifiedTokenName).append(":").append(new String(newTokenHash));
             }
-            assertBusy(() -> assertEquals("Waited too long for the updated file to be picked up", 0, latch.getCount()),
-                5, TimeUnit.SECONDS);
-            assertThat(store.getTokenHashes().get(qualifiedTokenName), equalTo(newTokenHash));
+            assertBusy(() -> {
+                assertThat("Waited too long for the updated file to be picked up", counter.get(), greaterThan(oldValue6));
+                assertThat(store.getTokenHashes().get(qualifiedTokenName), equalTo(newTokenHash));
+            }, 5, TimeUnit.SECONDS);
         }
     }
 

